---
layout:     post
title:      C++语言基础(9)——文件流
subtitle:   
date:       2020-03-20
author:     wukurua
header-img: img/C++/post-bg-C++.jpg
catalog: true
tags:
    - C++
---

# 一、异常处理 #



在编写程序时，应该考虑可以预料的运行异常，然后加入处理错误的代码，避免死机和灾难性的后果，而有正确合理的表现，从而提高程序的健壮性。

可以预料的运行异常有：

- 动态分配空间时可能不会成功
- 打开文件可能会失败
- 除法运算时分母可能为0
- 整数相乘可能溢出
- 指针可能越界……

# 二、C++异常处理机制 #
程序在产生错误后抛出异常，异常处理模块捕获并处理异常。
异常处理机制一般无法使程序恢复正常执行，即抛出异常之后，程序控制无法返回到抛出点。

	try
	{
	   	//出错时产生异常的代码
		//throw <表达式>;抛出任意类型的异常
		//异常抛出后，最近的一个匹配的异常处理器捕获该异常
	}
	catch（类型1  参数1）//捕捉异常
	{
	   	//针对类型1的异常处理
	}
	catch （类型2  参数2）
	{
	   	//针对类型2的异常处理
	}
	…
	catch （类型n  参数n）
	{
	   	//针对类型n的异常处理
	}

举例：

	int a,b;
	cin>>b>>a;
	int e;
	try
	{
		if(a==0)
		{
			throw e;
		}
	}
	catch(int e)
	{
		cout<<"除数为零!"<<endl;
	}

# 三、C++标准类的异常基类 #
exception 类是C++标准库函数抛出的所有异常的基类。
异常类反映的错误模型中错误被分成两个大类：

- 逻辑错误（logic error ）
- 运行时刻错误（run-time error）

![83D0OA.jpg](https://s1.ax1x.com/2020/03/15/83D0OA.jpg)

头文件：

	#include<exception>

使用：

	vector <int> intvec(10,1);
	try
	{
		intvec.at(20)=10;
	}
	catch(out_of_range &err)
	{
		cout<<err.what()<<endl;
	}


控制台：`invalid vector<T> subscript`